### DISCUSSION TITLE
Interpreting NaN values in IDP files for developing QC IDP thresholds

### DISCUSSION THREAD

**Discussion by dsodums** [McIntosh-Lab/tvb-ukbb#169] [Help]
I was examining the IDP values for the Rockland dataset outputs and was hoping to get more information about what the following variables represent in terms of NaNs and if this is of concern or not if we do have lines in the SC or TL that have NaNs. 

SC_nan_lines
SC_num_nan

TL_nan_lines
TL_num_nan

For example, line 205 is often an NaN in many of our outputs, I'm assuming this means that there might have been some problem with the structural processing such that ROI #205 is problematic. 

We are looking to develop thresholds to apply semi-automated QC so we have processed a subset of ~200 cases and manually QC'ed them, similar to what was done with the CAM-Can dataset in the tvb-ukbb paper, and I'm hoping to confirm this potential issue before we start running any analyses with the QC 1-5 indices and the IDP values.

**Comment by justinliuyuwang** [comment]
Hi Devin!

num_lines are the ROI #s of all-NaN lines
num_nan is the number of all-NaN lines

There may have been an error with tractography from/to ROI 205 if you are getting a NaN row. Depending on your end goals for the dataset, you may need to resolve this issue. If there are no obvious errors in structural and diffusion processing, then trying a different parcellation may be helpful.

**Reply by dsodums** [reply]
Hi Justin! 

Thanks so much. It looks like those cases do have some minor errors in tractography, my guess is that the most common ones are areas that are on the edge of the brain and might have been impacted by poor quality T1 extraction or impacted by motion. So we will either exclude those individuals from final analysis. 

In your group's experience, if there are just 1-2 ROIs that are commonly resulting in a NaN row, would you consider keeping the parcellation but editing it to remove just those ROIs?

If it's better to try a different parcellation, would it be correct to assume that going with a smaller number of ROIs could reduce the chance of this happening?

**Reply by dsodums** [reply]
Hi again,
After doing some research I'm thinking it might be best for us to go with a smaller parcellation such as the 220 Shaeffer-Tian one. Before test this out, I was hoping to check if there were NaN's in any of the actual FC matrices - I don't see this as one of the IDP's that is generated however - is it one of the set and I'm missing it, or do I need to check the actual fc.txt file/run a command that checks if there are NaNs in it?

**Reply by justinliuyuwang** [reply]
Hi Devin! Sorry for the late response. Yes a coarser parcellation like 220 would be a perfect next step. Unfortunately, we do not have an IDP for checking FC matrix NaNs at this point, sorry!

The code for printing any NaN-lines for an FC matrix in Python is as follows:


```python

    FC_path="path/to/fc.txt"
    nanlines=""
    with open(FC_path) as f:
        lines = f.read().splitlines()
        counter1=1
        counter=0
        for x in lines:

            x = x.split()

            nan_row=True
            for y in x:
                if y != "nan":
                    nan_row=False
            if nan_row == True:
                counter +=1
                nanlines = nanlines + str(counter1) + ", "
            counter1 +=1
    num_nanlines=counter

    print("number of nanlines: "+str(num_nanlines))
    print("nan lines: "+nanlines)    


```

**Reply by dsodums** [reply]
Hi Justin, 
No need to apologize at all! If anything, thank you for the help and support in such a short time. I ran the reparcellation with the 220 parcellation and am checking things out now. I'm guessing this Python script needs to be edited for each subject - editing the "path/to/fc.txt" and then ran for each subject of interest?

**Reply by justinliuyuwang** [reply]
Yes! Or if you are familiar enough with python, you can generate a python file using this script that takes in command line arguments so you can call it with `python FC_nan_counter.py ${subname}" within a loop that iterates through all your subjects.

**Reply by dsodums** [reply]
OK great - I am not very familiar with python actually, so I'll probably run it on a case by case basis for now. 

I'm actually dealing with one error though when running the code - I've pasted that code and created a .py file, and submitted it but I'm getting an unexpected indent error so I got rid of the extra spaces and it's now giving me the following error:


[soddev@cedar5 Rockland]$ python FC_Nan.py
Traceback (most recent call last):
  File "FC_Nan.py", line 21, in <module>
    print("number of nanlines: "+num_nanlines)
TypeError: can only concatenate str (not "int") to str



Here is the code I'm running:

FC_path="/home/soddev/scratch/Rockland/pipeline_outputs/reparcellate_SchaefferTian220/sub-A00028844/fMRI/rfMRI_0.ica/f>
nanlines=""
with open(FC_path) as f:
        lines = f.read().splitlines()
        counter1=1
        counter=0
        for x in lines:

            x = x.split()

            nan_row=True
            for y in x:
                if y != "nan":
                    nan_row=False
            if nan_row == True:
                counter +=1
                nanlines = nanlines + str(counter1) + ", "
            counter1 +=1
num_nanlines=counter

print("number of nanlines: "+num_nanlines)
print("nan lines: "+nanlines)

**Reply by justinliuyuwang** [reply]
ah yes, sorry about that - `print("number of nanlines: "+num_nanlines)` should be `print("number of nanlines: "+str(num_nanlines))`

i've adjusted the original script above to reflect that

**Reply by dsodums** [reply]
All good, thanks! I'm still getting an error though, with line 22:

 print("nan lines: "+nanlines)
        ^
SyntaxError: invalid syntax

**Reply by justinliuyuwang** [reply]
can you copy the exact script you're using please? i think you may have a typo somewhere in your code

**Reply by dsodums** [reply]
Thanks again for your help! Sure thing here it is:

FC_path="/home/soddev/scratch/Rockland/pipeline_outputs/reparcellate_SchaefferTian220/sub-A00028844/fMRI/rfMRI_0.ica/fc_TVBSchaeferTian220.txt"
nanlines=""
with open(FC_path) as f:
        lines = f.read().splitlines()
        counter1=1
        counter=0
        for x in lines:

            x = x.split()

            nan_row=True
            for y in x:
                if y != "nan":
                    nan_row=False
            if nan_row == True:
                counter +=1
                nanlines = nanlines + str(counter1) + ", "
            counter1 +=1
num_nanlines=counter

print("number of nanlines: "+str(num_nanlines)
print("nan lines: "+nanlines)

**Reply by justinliuyuwang** [reply]
You are missing a closing bracket after num_nanlines!

**Reply by dsodums** [reply]
Ah that's it, thanks so much!

In your experience, how concerned would your group be about a few NaNs in SC/TL if the FC matrices themselves aren't showing any NaNs?

**Reply by justinliuyuwang** [reply]
This is a difficult question to answer as it depends on your project needs. A few NaNs here and there may not be a huge point of concern, but how you treat nans in analysis is important. If you accept matrices with nans those nans should not be included in analysis.

**Reply by dsodums** [reply]
That makes total sense, many thanks! 

Would it be of interest to this group for us to post what we plan to do with the outputs in a separate discussion thread?

**Reply by noahfl** [reply]
Yeah absolutely! Any non-bug-related discussion about the pipeline, how people use it, approaches, etc. is more than welcome :slightly_smiling_face:
